### ISSUE TITLE
"Function not implemented" when trying to access a file within a container. Works fine from Windows. Need to restart Docker to fix.

### ISSUE THREAD

**Issue by bertrandgorge** [docker/for-win#5955]
- [x] I have tried with the latest version of my channel (Stable or Edge)
  - [x] I have uploaded Diagnostics
  - Diagnostics ID:8C13FE43-7A4A-460C-A391-AD73DC026E3C/20200306165543

Hello, I'm using Docker on Windows 10, and both my colleague and I start having a very frustrating behaviour with files suddenly stopping from being accessible... To fix I need to restart Docker.

### Expected behavior
The problem appears on all files that are mounted via a volume. When opening the file from Windows it works well.

### Actual behavior
In the container, when trying to read the file, I get : 

```
root@4bb61642e04e:/var/www/scripts# cat build_project.php
cat: build_project.php: Function not implemented

```
### Information

Please, help us understand the problem.  For instance:
  - Is it reproducible? --> yes, until your restart Docker desktop
  - Is the problem new? --> apparently yes, both my colleague and I started to have that yesterday. We've been working with this version of Docker for some time now.
  - Did the problem appear with an update? No, or maybe something from Windows 10 ?

  - Windows Version: 
  - Docker Desktop Version: Windows 10 Pro 1909 18363.657
  - Are you running inside a virtualized Windows e.g. on a cloud server or on a mac VM: no

**Comment by djs55** [comment]
Thanks for your report. Unfortunately I couldn't find anything useful in the logs and I don't know how to reproduce this myself. There are a number of places in the code where `ENOSYS` (function not implemented) is deliberately returned, because this is the signal for "I don't implement this FUSE function, please don't ask me again and do whatever it is another way". However at least one of these is clearly wrong, and there is no explicit logging. I've done 2 things:
1. I've added a ticket to our backlog to implement the rest of these functions
2. I've created a developer build with a log message in each place where I believe `ENOSYS` can come from: https://download-stage.docker.com/win/edge/43421/Docker%20Desktop%20Installer.exe

If you have time, could you reproduce the problem with this build and then upload fresh diagnostics? If I could narrow down the problem to the specific missing call then we could get a fix sooner.

Thanks for your help!

**Comment by bertrandgorge** [comment]
Hello David, I thought it was that kind of bug... I've downloaded your build and asked others in the team to do so too. We'll keep you updated if we manage to catch it once again!

**Comment by spipu** [comment]
i have the same pb :(

**Comment by djs55** [comment]
@spipu could you try the build from [this comment](https://github.com/docker/for-win/issues/5955#issuecomment-597684110) and then when it reproduces, upload a diagnostics and report the id here? Thanks in advance!

**Comment by bertrandgorge** [comment]
(not a single case on my side since then unfortunately)

**Comment by spipu** [comment]
BEE314D9-AEF0-4EFA-A7F5-7FCA9AAA68F9/20200318172344

**Comment by djs55** [comment]
@spipu thanks for the diagnostics. Could you tell me the filename of the files (or directories) which failed?

**Comment by spipu** [comment]
it was /var/www/spareka/website/index.php

**Comment by spipu** [comment]
when i start the container, it works during 10 seconds. then the container restart and i have the error.
any file that is coming from the mount directory from windows has the error.

**Comment by alex-shchetkov** [comment]
I had a similar issue today and finally figured out the cause in my case. 

(in python) I tried importing from a filepath that was longer than 260 characters. 
Everything breaks afterward, and all commands return "Function not implemented" until Docker is completely restarted

**Comment by sorbaugh** [comment]
Hello,

I started having the same problem 3 days ago. I am using Docker for Windows 2.2.0.3. I've been using this version of Docker for a while with absolutely no problems before. Problem >might< have started after a updating Windows.

One interesting thing I've observed while trying to debug this mess, is that if I attach to the container after a fresh restart of Docker, I can read the mounted files just fine (using vim for example). But after I open my website in the browser once (which fails with a 403 Forbidden), trying to read any mounted files ends with a Reading Error.

Container Logs show:
[Mon Mar 23 10:41:39.453490 2020] [core:error] [pid 18] (38)Function not implemented: [client xxx.xxx.xxx.xxx:46170] AH00036: access to /index.php failed (filesystem path '/var/www/html/index.php')

Using vim within container (file is shown empty, applies to any mounted file):
"index.php" [READ ERRORS] 0L, 0C

**Comment by arnesacnussem** [comment]
A957D1CA-D8EE-43F0-8F02-685804769BE3/20200405062145

this happend to me after moving some file to a sub-dir
restart docker fixs this for now

```
/app # cat package.json
cat: read error: Function not implemented
```

**Comment by richardjlyon** [comment]
Symptom for me was 'Bus error' (hopefully others googling the same issue will find this), turns out php was trying to eval a file in a mounted dir, which was returning 'Function not implemented'

**Comment by djs55** [comment]
Thanks for the report and your patience. I've found and fixed one possible source of this problem and have a candidate build here: https://download-stage.docker.com/win/edge/44390/Docker%20Desktop%20Installer.exe . If you have time to test, could you let me know if it's any better? If not, could you upload a fresh set of diagnostics? Thanks!

**Comment by arnesacnussem** [comment]
@djs55 

doesn't fix this
A957D1CA-D8EE-43F0-8F02-685804769BE3/20200417155633
![image](https://user-images.githubusercontent.com/9441546/79589419-f1d14780-8107-11ea-8e0e-ac17e26769b9.png)

**Comment by WaveHack** [comment]
Can confirm same issue. Docker Edge v19.03.8.

Happened to me randomly where PHP's `composer install` complained about "file_get_contents(): read of 8192 bytes failed with errno=38 Function not implemented". (putting this here in case somebody Googles it).

After finding this thread I confirmed that the underlying error was with `cat`. Reportedly restarting the Docker client resolved this issue for me.

**Comment by desliner** [comment]
In my case Docker restart didn't help and the issue always persisted.
The workaround was to downgrade all the way to stable-2.1.0.5.40693 version.

**Comment by jeremynikolic** [comment]
> 
> 
> Can confirm same issue. Docker Edge 2.3.0.0 (44472), Docker Engine v19.03.8.
> 
> Happened to me randomly where PHP's `composer install` complained about "file_get_contents(): read of 8192 bytes failed with errno=38 Function not implemented". (putting this here in case somebody Googles it). After finding this thread I confirmed that the underlying error was with `cat`.
> 
> As reported restarting the Docker client resolved this issue for me. Unfortunately I do not know what caused it, and I haven't been able to reproduce it since. In case I do I'll report back here.

Same case here. Docker Desktop 2.3.0.2 and Docker Engine 19.03.8 
I started having issue loading my site pages with an "fwrite exception" then encountered the composer "fread".
In my case I had to remove container/image/volumes, restart the client and rebuild it all for it to work again.
A first attempt at rebuilding would not work has fread error happened during the containers building process (nginx and mysql in this case were trying to read conf files which are mapped from my windows host)

**Comment by djs55** [comment]
@jeremynikolic thanks for your report. Could you reproduce the issue and then upload a diagnostics report? The diagnostics are in the whale menu -> Troubleshoot. When the report has been created, please upload it and quote the id number here. This will allow me to take a look at the logs. Thanks in advance!

**Comment by jeremynikolic** [comment]
@djs55 I finally fixed it so I am not able to replicate it now.

But here what I could understand of the problem:

Prior to the issue I updated windows 10 and docker desktop afterward. 

For some reason Doker Desktop was not running as administrator afterwards, changing the it to run as Administrator fixed the issue for me. 

The error was happening when my container was trying to read or write files. For example: mysql or nginx which are mapped to my host. 
So it seems that windows was "simply" refusing access to my drive as the docker engine was not running as administrator.

**Comment by djs55** [comment]
@jeremynikolic that's very interesting, thanks. Perhaps there's a problem with handling permission errors which you worked around by using `Administrator`. That's a good lead!

**Comment by Gitsack** [comment]
Same here, when running Docker Windows as administrator, the error is gone.

**Comment by jeremynikolic** [comment]
@djs55 
The cause being the windows update seems more plausible because I am now getting other applications throwing errors regarding read/write permissions now.
Even just downloading a file in the browser is blocked where it wasn't before...

**Comment by mattisdada** [comment]
I'm going to be very unhelpful as well, as this didn't happen on my machine but someone elses (so I don't have access to logs). 
It was also after a `composer install` and it started giving this error message. Restarting Docker as recommended here fixed the issue.

**Comment by treestonemedia** [comment]
restarting docker worked for me

**Comment by SerethiX** [comment]
I had the very same issue:

```
webserver_1  | 2020/05/18 15:42:14 [crit] 1#1: pread() "/etc/nginx/conf.d/default.conf" failed (38: Function not implemented)
webserver_1  | nginx: [crit] pread() "/etc/nginx/conf.d/default.conf" failed (38: Function not implemented)
webserver_1  | 2020/05/18 15:43:20 [crit] 1#1: pread() "/etc/nginx/conf.d/default.conf" failed (38: Function not implemented)
webserver_1  | nginx: [crit] pread() "/etc/nginx/conf.d/default.conf" failed (38: Function not implemented)
webserver_1  | 2020/05/18 15:47:09 [crit] 1#1: pread() "/etc/nginx/conf.d/default.conf" failed (38: Function not implemented)
webserver_1  | nginx: [crit] pread() "/etc/nginx/conf.d/default.conf" failed (38: Function not implemented)
```
I restarted Windows, after that it worked again. Yesterday there were no issues using the software.

**Comment by JulienQueinnec** [comment]
I have the same issue.
Friday it was working but not today.

**Comment by qwergram** [comment]
Contributing a potential work around. I get this same exact error when I `docker-compose up` where this `docker-compose.yml` triggered the issue. 
```yml
version: '2'

services:
  redis:
    restart: always
    image: sameersbn/redis:4.0.9-2
    command:
    - --loglevel warning
    volumes:
    - ./redis:/var/lib/redis:Z
```

However, by changing the following lines I was able to get it going.

Changing:
```yml
# Line 10-11
volumes:
 - redis:/var/lib/redis:Z
```
and adding 
```yml
volumes:
  redis:
```
at the bottom seemed to fix it.

**Comment by SerethiX** [comment]
Update: it's broken again.

Nothing changed on my side, container was running fine when suddenly yarn stopped working, after a restart the build failes

```
yarn_1       | yarn install v1.22.4
yarn_1       | Error: ENOSYS: function not implemented, read
yarn_1       |     at Object.readSync (fs.js:524:3)
yarn_1       |     at tryReadSync (fs.js:349:20)
yarn_1       |     at Object.readFileSync (fs.js:378:19)
yarn_1       |     at onUnexpectedError (/opt/yarn-v1.22.4/lib/cli.js:92560:100)
yarn_1       |     at /opt/yarn-v1.22.4/lib/cli.js:92672:9
web_yarn_1 exited with code 1
```

Diagnose ID: A0851DC0-C4BD-4834-A013-B1D7B44C767A/20200519215959

**Comment by Julien-e** [comment]
Hello,

Same error on nginx/php7 with symfony.

```
php_1            | [20-May-2020 07:41:47] WARNING: [pool symfony] child 7 exited on signal 7 (SIGBUS) after 115.291748 seconds from start
php_1            | [20-May-2020 07:41:47] NOTICE: [pool symfony] child 13 started
php_1            | [20-May-2020 07:41:47] WARNING: [pool symfony] child 8 exited on signal 7 (SIGBUS) after 115.307269 seconds from start
php_1            | [20-May-2020 07:41:47] NOTICE: [pool symfony] child 14 started
php_1            | [20-May-2020 07:41:47] WARNING: [pool symfony] child 11 exited on signal 7 (SIGBUS) after 76.331625 seconds from start
php_1            | [20-May-2020 07:41:47] NOTICE: [pool symfony] child 15 started
php_1            | [20-May-2020 07:41:47] WARNING: [pool symfony] child 12 exited on signal 7 (SIGBUS) after 0.307129 seconds from start
php_1            | [20-May-2020 07:41:47] NOTICE: [pool symfony] child 16 started
```

**Comment by Syntaf** [comment]
Ran into this issue while renaming some react components on my Rails 6 project. I'm using hot module reloading and running webpack in a separate container; at one point the webpack container failed to compile and all containers sharing the same volume started erroring with stuff like

```
 /usr/local/lib/ruby/2.7.0/rubygems/bundler_version_finder.rb:101:in `read': Function not implemented @ io_fread - /opt/app/myapp/Gemfile.lock (Errno::ENOSYS)
```

I ran bash on the webpacker container and noticed that basic commands also were failing (like mentioned above):

```
user@4c3efb4662cd:/opt/app/myapp$ cat Gemfile.lock
cat: Gemfile.lock: Function not implemented
```

I managed to get things working again by restarting Docker Desktop. Possibly something is awry when mulitple containers share the same volume

**Comment by iamalbert** [comment]
I got sample problem here. Cannot read or import any file in container if I git-checkout to another branch.
The only solution is to restart docker completely.

#6337 is fixed and closed but this problem still exists.

**Comment by cicorias** [comment]
I've been able to repro this by doing a `npx rimraf ./node_modules` as an `initializeCommand`, then for a `postCreateCommand` of `npm ci` --- where an existing `package-log.json` or `npm-shrinkwrap.json` exists.

It seems the file io is overwhelming enough that it mucks up the ssh connections minimum, then I see `ENOSYS: function not implemented, read` when running a simple `npm i` in the workspace.

**Comment by nedkelly** [comment]
This issue is getting progressively worse, previously like others have mentioned it occurs during `yarn install`, `npm install`  operations and the only way to fix it has been to restart Docker Desktop, but now even during the middle of my development build in a NextJS project it suddenly kills my builds this is preventing me from being able to develop.

![screenshot-nimbus-capture-2020 05 28-11_25_11](https://user-images.githubusercontent.com/569534/83089344-75a02b80-a0d9-11ea-9d7c-cff6cf904acf.png)

It's not possible to resume development after this occurs so I have to stop what I'm doing restart Docker Desktop and hope it doesn't happen again 🤞 

![screenshot-nimbus-capture-2020 05 28-11_39_50](https://user-images.githubusercontent.com/569534/83089426-a1bbac80-a0d9-11ea-903d-7ad52cf4f4d0.png)

Docker Desktop:
version 2.3.0.3 (45519)
Engine 19.03.8

Windows 10 Pro Version	10.0.18362 Build 18362

**Comment by TheKeymaster** [comment]
Having the same issue. I am using Docker Windows 2.3.0.3 and it still does not seem to be fixed.
That error prevents me from developing cool things, and isn't always fixable by "just restarting docker".

Please fix this bug asap, as this bug actually costs so much time, and docker is basically unusable when it occurs.

**Comment by desliner** [comment]
Feel free to downgrade to the [stable-2.1.0.5.40693](https://docs.docker.com/docker-for-windows/release-notes/#docker-desktop-community-2105) version to work around this issue.
IIUC it has previous implementation of file sharing, which doesn't seem to be affected by the bug.

**Comment by iamalbert** [comment]
2.2.0.5 works too

**Comment by kheera** [comment]
I think I have the same issues under a win10/docker (2.3.03 (45519)) shared volume, restarting docker seems to fix it for a short while. I have no idea what's triggering the issue. In my case docker is running an apache/laravel combo.

root@96f37b67e485:/var/www# cat test.txt
cat: test.txt: Function not implemented

root@96f37b67e485:/var/www# php put-test.php
Bus error

root@96f37b67e485:/var/www# cp test.txt /root
cp: error reading 'test.txt': Function not implemented

**Comment by redfellow** [comment]
I'm also affected on latest stable

**Comment by djs55** [comment]
Thanks all for your reports. I've found a potential cause of the "function not implemented" error in the  file sharing service. I've created an experimental build based on the latest stable. If you'd like to try it the build is here: https://desktop-stage.docker.com/win/stable/45688/Docker%20Desktop%20Installer.exe . Let me know if it works for you, and if not, please describe what you see, upload diagnostics and quote the id here. Thanks again!

**Comment by rfay** [comment]
Demonstration case: Windows 2004, Docker 2.3.0.3, Hyper-V mode, git-bash:

```
docker run -v "//c/Users:/junkmount"  --rm busybox sh -c "df /junkmount"
df: /junkmount: Function not implemented
Filesystem           1K-blocks      Used Available Use% Mounted on
```

**Comment by jhemarcos** [comment]
The following problem happened to me after deleting a container file in the windows host, while the container was running with docker-compose with a mounted volume:

`nginx: [crit] pread() failed (38: Function not implemented)`

Restart Docker for Windows solved the problem.

**Comment by yoieh** [comment]
started to get the samme error as @jhemarcos some days ago after updating docker for windows and its the second timme now.

> nginx: [crit] pread() failed (38: Function not implemented)

but

> Restart Docker for Windows solved the problem.

Had to restarting it to times to resolve my issue.

docker-compose.yml
```YAML
webserver:
    image: nginx:stable-alpine
    container_name: lo_nginx
    restart: unless-stopped
    tty: true
    volumes:
      - .:/var/www/html
      - ./docker/nginx/nginx.conf:/etc/nginx/conf.d/default.conf
    ports:
      - "${WEBSERVER_PORT}:80"
    networks:
      - lo-network
```

./docker/nginx/nginx.conf
```Nginx
server {
    listen 80;

    root /var/www/html/public;

    index index.php index.htm index.html;

    charset utf-8;

    location / {
        try_files $uri $uri/ /index.php?$query_string;
    }

    location = /favicon.ico { access_log off; log_not_found off; }
    location = /robots.txt  { access_log off; log_not_found off; }

    client_max_body_size 108M;

    access_log off;
    error_log  /var/log/nginx/application.error.log error;

    location ~ \.php$ {
        # php-fpm container url
        fastcgi_pass app:9000;
        fastcgi_index index.php;
        fastcgi_param SCRIPT_FILENAME $document_root$fastcgi_script_name;
        fastcgi_buffers 16 16k;
        fastcgi_buffer_size 32k;
        include fastcgi_params;
    }

    location ~ /\.ht {
        deny all;
    }
}
```

![image](https://user-images.githubusercontent.com/5743837/84143141-577de800-aa56-11ea-98c0-4d737c3316e9.png)

**Comment by djs55** [comment]
@rfay thanks for the example, it also reproduces for me on that build. I think this instance of the bug is in the handling of the `//c/Users` path. I've got yet another development build if anyone would like to try it: https://desktop-stage.docker.com/win/edge/45823/Docker%20Desktop%20Installer.exe . When I install this build the command works as expected: (after I'm prompted to allow the directory to be shared)
```
> docker run -v "//c/Users:/junkmount"  --rm busybox sh -c "df /junkmount"
Filesystem           1K-blocks      Used Available Use% Mounted on
grpcfuse             486931040 119100336 367830704  24% /junkmount
```
Thanks again!

**Comment by mdecurtins** [comment]
Same problem, using 2.3.0.3. Been experiencing this for at least a month now.

I usually see it if I've switched between branches, thus changing what files are seen in the mounted volume. Restarting Docker does seem to work, for now.

**Comment by andrewwatkin** [comment]
Same problem using 2.3.0.3. I think that the problem is load dependent - it was fine running an existing container until I had to rebuild one which involved doing "npm install @angular/cli". Then lots of ENOTENT and ENOSYS errors and the container just hanging forever - 1% CPU in docker stats.

Windows 10. Container sharing a volume with Windows. The share contains node_modules/ which is obviously hit hard by npm.

I have reverted to 2.1.0.5 and it all works fine.

**Comment by diogoaguiar** [comment]
Had the same problem on 2.3.0.3. Downgraded to 2.2.0.5 and it has been working fine so far.

**Comment by viraj892** [comment]
Same issue with 2.3.0.3, multiple restart were required to get my containers up and running. I haven't tried downgrading to 2.2.05 though, will be doing that as soon as I encounter the issue again.

**Comment by stephen-turner** [comment]
This is fixed in the latest Edge release, 2.3.2.0.

**Comment by KavishkaRajakaruna** [comment]
Restart docker if that didn't work out try restarting the computer. This is an issue with the file system mapping among the OS & Docker service

**Comment by djs55** [comment]
The fix from edge 2.3.2.0 has now been released in [stable 2.4.0.0](https://desktop.docker.com/win/stable/48506/Docker%20Desktop%20Installer.exe).
